### PR TITLE
Adds greyscale color selection to vending machines

### PR DESCRIPTION
**Pull request by ninjanomnom**
## About The Pull Request

This allows any item vended from a vending machine to potentially be color edited before dispensing. This only works with items that have been converted to the GAGS system so the selection is limited for now.

https://user-images.githubusercontent.com/1234602/117590213-4e7d1700-b0fc-11eb-9360-ce0bb0a4e844.mp4

Yes this is blatant bait to get more things converted.

While working on this I fixed item greyscale updates to include their held and worn states, the vending machine ui is now tsx instead of js, icons generated by gags are created with an error state by default, the greyscale color menu defaults to not show you the full debug preview, and the wording in the debug menu is a bit more in-character friendly.

## Changelog
:cl:
add: Greyscale items using GAGS in vending machines can have their colors chosen before vending. Not many things are capable of this yet but expect more to come.
/:cl:

### CODE REVIEW

**Comment by Fikou** [comment]
can you make these more player facing rather than showing /datum typepaths for crew?

**Comment by ninjanomnom** [comment]
Those configs don't really *have* names given to them

**Comment by Fikou** [comment]
just add var/name

**Comment by ninjanomnom** [comment]
I'll have to think about it, I'd prefer those objects stay as absolute minimal as possible and I'm not sure adding a name there is worth it

**Comment by tralezab** [comment]
I agree with fikou that showing byond paths in game is pretty lame

File: tgui/packages/tgui/interfaces/GreyscaleModifyMenu.tsx [hunk 1]
```diff
@@ -107,27 +109,29 @@ const ColorDisplay = (props, context) => {
...
-      </Flex>
-      <Flex mx="25%">
-        <SingleDirection dir={Direction.SouthWest} />
-        <SingleDirection dir={Direction.South} />
-        <SingleDirection dir={Direction.SouthEast} />
-      </Flex>
-    </Section>
+    <Box grow>
+      <Stack vertical>
+        <Flex grow fluid>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/GreyscaleModifyMenu.tsx:114]: > `<Flex grow>`

Flex doesn't work like that, you know.

File: tgui/packages/tgui/interfaces/Vending.tsx [hunk 2]
```diff
@@ -1,32 +1,85 @@
-import { classes } from 'common/react';
+import { classes } from '../../common/react';
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Vending.tsx:1]: Revert this, you are not supposed to escape package boundaries

File: tgui/packages/tgui/interfaces/Vending.tsx [hunk 3]
```diff
@@ -1,32 +1,85 @@
...
 import { Box, Button, Section, Table } from '../components';
 import { Window } from '../layouts';
 
+type VendingData = {
+  onstation: boolean
+  department: string
+  jobDiscount: number
+  product_records: Array<ProductRecord>
+  coin_records: Array<CoinRecord>
+  hidden_records: Array<HiddenRecord>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Vending.tsx:12]:
```
ProductRecord[]
CoinRecord[]
HiddenRecord[]
```

File: tgui/packages/tgui/interfaces/GreyscaleModifyMenu.tsx [hunk 4]
```diff
@@ -187,20 +210,27 @@ export const GreyscaleModifyMenu = (props, context) => {
...
-        <Button
-          content="Refresh Icon File"
-          onClick={() => act("refresh_file")}
-        />
-        {" "}
+        {
+          data.generate_full_preview
+            ? <Button content="Refresh Icon File" onClick={() => act("refresh_file")} />
+            : ""
+        }
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/GreyscaleModifyMenu.tsx:223]: Refer to what I posted on Discord.

File: tgui/packages/tgui/interfaces/Vending.tsx [hunk 5]
```diff
@@ -1,32 +1,85 @@
...
     product,
     productStock,
     custom,
   } = props;
-  const {
-    onstation,
-    department,
-    user,
-    jobDiscount,
-  } = data;
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Vending.tsx:18]: But why

File: tgui/packages/tgui/interfaces/Vending.tsx (hunk 2 above)
Comment (ninjanomnom) [reply tgui/packages/tgui/interfaces/Vending.tsx:1]: The code was refusing to build in tsx without this

File: tgui/packages/tgui/interfaces/GreyscaleModifyMenu.tsx (hunk 1 above)
Comment (ninjanomnom) [reply tgui/packages/tgui/interfaces/GreyscaleModifyMenu.tsx:114]: Ah yeah I forgot to come back here and clean stuff up after getting it to finally work. Was having issues when dev server just suddenly stopped updating uis and no changes were taking effect.

File: tgui/packages/tgui/interfaces/Vending.tsx (hunk 5 above)
Comment (ninjanomnom) [reply tgui/packages/tgui/interfaces/Vending.tsx:18]: These were counting as any typed and specifying the types here a second time seemed wasteful when it's trivial to just do `data.user` and such

**Comment by optimumtact** [comment]
have you considered the impact of all this icon blending as a tradeoff for rsc size, which while high, is not the majority of our bandwidth costs?

Also what danger is there in allowing end user clients to potentially spam icon operations via this menu?

File: tgui/packages/tgui/interfaces/Vending.tsx (hunk 5 above)
Comment (stylemistake) [reply tgui/packages/tgui/interfaces/Vending.tsx:18]: > These were counting as any typed

It's not possible, if these are any typed, then the whole data is any typed.

**Comment by ninjanomnom** [comment]
The benefit to me for greyscale isn't primarily in reducing rsc size, that's just a nice side effect that's the easiest to measure. I see it as more useful for its flexibility in usecases like allowing recolors mid round as done here. The processing costs are fairly minimal but yeah I probably should put some additional protection in place to avoid people intentionally bloating the memory costs.

**Comment by cacogen** [comment]
This is really cool

**Comment by ninjanomnom** [comment]
@optimumtact I've made it so the menu limits to one refresh per second at most and the preview in the menu doesn't save anything unless the colors/config actually get applied to something. So the worst that can happen is generating a single greyscale sprite per second, and maybe the client crashing themselves due to byond's client side caching.

**Review by Twaticus — APPROVED** [review]

**Comment by maxymax13** [comment]
is this the PR that put the funny e on everything

**Comment by optimumtact** [comment]
yes

**Comment by cacogen** [comment]
worth the cost

**Pull request merged**